ngrinder-controller/src/main/java/org/ngrinder/agent/service/AgentService.java

```diff
@@ -0,0 +1,489 @@
+		}
+		agentManager.runAgent(singleConsole, grinderProperties, necessaryAgents);
+	}
+
+	/**
+	 * Select agent. This method return agent set which is belong to the given user first and then share agent set.
+	 *
+	 * @param user          user
+	 * @param allFreeAgents agents
+	 * @param agentCount    number of agent
+	 * @return selected agent.
+	 */
+	private Set<AgentInfo> selectAgent(User user, Set<AgentInfo> allFreeAgents, int agentCount) {
+		Stream<AgentInfo> ownedFreeAgentStream = allFreeAgents.stream().filter(isOwnedAgent.apply(user.getUserId()));
+		Stream<AgentInfo> freeAgentStream = allFreeAgents.stream().filter(isCommonAgent);
+		return concat(ownedFreeAgentStream, freeAgentStream).limit(agentCount).collect(toSet());
+	}
+
+	/**
+	 * Stop agent. If it's in cluster mode, it queue to agentRequestCache.
+	 * otherwise, it send stop message to the agent.
+	 *
+	 * @param ip ip of agent to stop.
+	 * @param name host name of agent to stop.
+	 */
+	@Override
+	public void stop(String ip, String name) {
+		AgentInfo agentInfo = agentInfoStore.getAgentInfo(createAgentKey(ip, name));
+		if (agentInfo == null) {
+			return;
+		}
+		publishTopic(agentInfo, STOP_AGENT);
+	}
+
+	public void stop(AgentControllerIdentityImplementation agentIdentity) {
+		agentManager.stopAgent(agentIdentity);
+	}
+
+
+	@Override
+	public void update(String ip, String name) {
+		AgentInfo agentInfo = agentInfoStore.getAgentInfo(createAgentKey(ip, name));
+		if (agentInfo == null) {
+			return;
+		}
+		publishTopic(agentInfo, UPDATE_AGENT);
+	}
+
+	@Override
+	public SystemDataModel getSystemDataModel(String ip, String name, String region) {
+		AgentControllerIdentityImplementation agentIdentity = getAgentIdentityByIpAndName(ip, name);
+		return agentIdentity != null ? agentManager.getSystemDataModel(agentIdentity) : new SystemDataModel();
+	}
+
+	/**
+	 * Update the agent
+	 *
+	 * @param agentIdentity agent identity to be updated.
+	 */
+	public void updateAgent(AgentIdentity agentIdentity) {
+		agentManager.updateAgent(agentIdentity, agentManager.getAgentForceUpdate() ? "99.99" : config.getVersion());
+	}
+
+
+	private String createAgentKey(String ip, String name) {
+		return ip + "_" + name;
+	}
+
+	/**
+	 * Ready agent state count return
+	 *
+	 * @param user The login user
+	 * @param targetRegion targetRegion The name of target region
+	 * @return ready Agent count
+	 */
+	@Override
+	public int getReadyAgentCount(User user, String targetRegion) {
+		int readyAgentCnt = 0;
+		String myOwnAgent = targetRegion + "_owned_" + user.getUserId();
+		for (AgentInfo agentInfo : getAllReady()) {
+			if (!agentInfo.isApproved()) {
+				continue;
+			}
+			String fullRegion = agentInfo.getRegion();
+			if (StringUtils.equals(fullRegion, targetRegion) || StringUtils.equals(fullRegion, myOwnAgent)) {
+				readyAgentCnt++;
+			}
+		}
+		return readyAgentCnt;
+	}
+
+	@Override
+	public List<AgentInfo> getAllVisible() {
+		return getAllActive();
+	}
```

Comment:
What about just delete out ```getAllVisible```

Reply:
It will be extract to abstract class for backward compatibility.